### PR TITLE
Gift Entry: Initial Commit for creating JWTs programmatically

### PR DESCRIPTION
**Pull request by voduyemi**
# Critical Changes

# Changes
- Utility for creating JWTs for payment/purchase authorization on the elevate platform.

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by voduyemi** [comment]
**lurch: attach W-039948

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HtAyQAK" target="_blank">W-039948</a>

**Comment by voduyemi** [comment]
**lurch: attach W-039948

File: src/classes/UTIL_Jwt.cls [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+    /**
+     * @description helper method to sign jwt using payload as string and base64 format key
+     * @param   String payload, jwt payload
+     * @param   String key, base64 encoded key
+     * @return  String, returns the compact serialization representation of the jwt as a concatenated string,
+     *          with the encoded jwt header, encoded jwt payload, and encoded jwt signature strings
+     *          separated by period ('.') characters.
+     */
+    public static String getSignedJWTFromString(String payload, String key) {
```
Comment (vmartinov-sforg) [inline src/classes/UTIL_Jwt.cls]: I would like to see detailed in the comments an example of the payload with some claims in it.

File: src/labels/CustomLabels.labels [hunk 2]
```diff
@@ -4599,6 +4599,22 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
         <shortDescription>Common label for General Accounting Unit</shortDescription>
         <value>General Account Unit</value>
     </labels>
+    <labels>
+        <fullName>commonJWTInvalidKey</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure message if an invalid key has been provided</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels]:
```suggestion
        <shortDescription>Failure message if an invalid key was provided</shortDescription>
```

File: src/labels/CustomLabels.labels [hunk 3]
```diff
@@ -4599,6 +4599,22 @@ To check your GAU Allocation settings, go to Donations | GAU Allocations.</value
...
+        <value>The key must contain a value and use the base64 format.</value>
+    </labels>
+    <labels>
+        <fullName>commonJWTInvalidPayload</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure message if an invalid JSON payload has been provided</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels]:
```suggestion
        <shortDescription>Failure message if an invalid JSON payload was provided</shortDescription>
```

File: src/labels/CustomLabels.labels [hunk 4]
```diff
@@ -5994,12 +6018,20 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
...
     </labels>
     <labels>
-        <fullName>geHelpTextHouseholdMapping</fullName>
+        <fullName>geJWTUtilInvalidBase64Message</fullName>
         <categories>Gift Entry</categories>
         <language>en_US</language>
         <protected>true</protected>
-        <shortDescription>Help text for Household Object Group</shortDescription>
-        <value>Additional information about Contact1's Household Account.</value>
+        <shortDescription>Failure message if an invalid base64 string has been provided</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels]:
```suggestion
        <shortDescription>Failure message if an invalid base64 string was provided</shortDescription>
```

File: src/labels/CustomLabels.labels [hunk 5]
```diff
@@ -5994,12 +6018,20 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
...
+        <value>The base64 format is invalid. The number of characters must be a multiple of 4 and can only contain these characters: [a-z A-Z 0-9 + / =]</value>
+    </labels>
+    <labels>
+        <fullName>geJWTUtilInvalidBase64UrlMessage</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure message if an invalid base64 url has been provided</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels]:
```suggestion
        <shortDescription>Failure message if an invalid base64 url was provided</shortDescription>
```

File: src/labels/CustomLabels.labels [hunk 6]
```diff
@@ -6181,6 +6213,46 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <!-- {0} is a string. Name of the template that was created. -->
         <value>The "{0}" template was updated.</value>
     </labels>
+    <labels>
+        <fullName>geUtilJWTTestEmptyValue</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Unit test error: test with empty value</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels]: @ptewson-sfdo @vmartinov-sforg @psadie-sfdo I defer to you all for clear verbiage around the unit test labels.

Comment (melissabarber) [reply src/labels/CustomLabels.labels]: Please don a hat of a brand new hire who runs into these unit tests and make sure the person would be able to know what to check if they need to fix anything.

**Comment by melissabarber** [comment]
@voduyemi I added some minor changes to the custom labels to tighten up the language a little. Sorry I didn't catch those when working through them in the spreadsheet. Let me know if you have any questions.

@ptewson-sfdo @vmarasow @psadie-sfdo @kenlewis-sfdo I'm all signed off on this PR. Thanks!

File: src/labels/CustomLabels.labels (hunk 6 above)
Comment (ptewson-sfdo) [reply src/labels/CustomLabels.labels]: @melissabarber I don't think this label is actually being used - @voduyemi should we remove it? Or did I just search wrong?

File: src/labels/CustomLabels.labels [hunk 7]
```diff
@@ -6181,6 +6213,46 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
...
+    <labels>
+        <fullName>geUtilJWTTestInvalidString</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Unit test error: An invalid string was provided but no exception was thrown</shortDescription>
+        <value>An invalid string was provided but no exception was thrown</value>
+    </labels>
+    <labels>
+        <fullName>geUtilJWTTestNullValue</fullName>
```
Comment (ptewson-sfdo) [inline src/labels/CustomLabels.labels]: Not sure if this label is currently being used either?

File: src/package.xml [hunk 8]
```diff
@@ -565,6 +565,8 @@
         <members>UTIL_IntegrationGateway_TEST</members>
         <members>UTIL_JobProgress_CTRL</members>
         <members>UTIL_JobProgress_CTRL_TEST</members>
+        <members>UTIL_Jwt</members>
```
Comment (ptewson-sfdo) [inline src/package.xml]: @voduyemi I think all those new labels need to be included in package.xml!!

File: src/labels/CustomLabels.labels (hunk 6 above)
Comment (voduyemi) [reply src/labels/CustomLabels.labels]: @ptewson-sfdo You searched right! I was supposed to remove it.

File: src/package.xml (hunk 8 above)
Comment (voduyemi) [reply src/package.xml]: @ptewson-sfdo They've been included.

**Pull request merged**